src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,19 +1357,90 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Optional base class for <see cref="IValidateSetValuesGenerator"/> implementations that want a default implementation to cache valid values.
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Initializes a new instance of the CachedValidValuesGeneratorBase class.
+        /// </summary>
+        /// <param name="cacheExpirationInSeconds">Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.</param>
+        public CachedValidValuesGeneratorBase(int cacheExpirationInSeconds)
+        {
+            _validValuesCacheExpiration = cacheExpirationInSeconds;
+        }
+
+        private int _validValuesCacheExpiration;
+
+        /// <summary>
+        /// Abstract method to generate a valid values.
+        /// </summary>
+        public abstract string[] GenerateValidValues();
+
+        /// <summary>
+        /// Get a valid values.
+        /// </summary>
+        public string[] GetValidValues()
+        {
+            // Because we have a background task to clear the cache by '_validValues = null'
+            // we use the local variable to exclude a race condition.
+            var validValuesLocal = _validValues;
+            if (validValuesLocal != null)
+            {
+                return validValuesLocal;
+            }
+
+            var validValuesNoCache = GenerateValidValues();
+
+            if (validValuesNoCache == null)
+            {
+                throw new ValidationMetadataException(
+                    "ValidateSetGeneratedValidValuesListIsNull",
+                    null,
+                    Metadata.ValidateSetGeneratedValidValuesListIsNull);
+            }
+
+            if (_validValuesCacheExpiration > 0)
+            {
+                Task.Delay(_validValuesCacheExpiration * 1000).ContinueWith((task) => _validValues = null);
```

Comment:
I swaped the lines.

Please clarify. If you mean that the race condition is related with clearing `_validValues` - we add a protect above (validValuesLocal) and return validValuesNoCache below - so clearing `_validValues` should not cause problems.